crates/subspace-service/src/sync_from_dsn/snap_sync.rs

```diff
@@ -95,55 +97,85 @@ pub(crate) async fn snap_sync<Backend, Block, AS, Client, PG, NR>(
+// Get blocks from the last segment or from the segment containing the target block.
+// Returns encoded blocks collection and used segment index.
+pub(crate) async fn get_blocks_from_target_segment<AS, PG>(
     segment_headers_store: &SegmentHeadersStore<AS>,
     node: &Node,
     piece_getter: &PG,
-    fork_id: Option<&str>,
-    client: &Arc<Client>,
-    import_queue_service: &mut IQS,
-    network_request: &NR,
-    sync_service: &SyncingService<Block>,
-) -> Result<(), Error>
+    target_block: Option<BlockNumber>,
+) -> Result<Option<(SegmentIndex, VecDeque<(BlockNumber, Vec<u8>)>)>, Error>
 where
-    B: sc_client_api::Backend<Block>,
-    PG: DsnSyncPieceGetter,
     AS: AuxStore,
-    Block: BlockT,
-    Client: HeaderBackend<Block>
-        + ClientExt<Block, B>
-        + ProvideRuntimeApi<Block>
-        + ProofProvider<Block>
-        + LockImportRun<Block, B>
-        + Send
-        + Sync
-        + 'static,
-    Client::Api: SubspaceApi<Block, FarmerPublicKey> + ObjectsApi<Block>,
-    IQS: ImportQueueService<Block> + ?Sized,
-    NR: NetworkRequest,
+    PG: DsnSyncPieceGetter,
 {
-    debug!("Starting snap sync...");
-
     sync_segment_headers(segment_headers_store, node)
         .await
         .map_err(|error| format!("Failed to sync segment headers: {}", error))?;
 
-    let last_segment_index = segment_headers_store
-        .max_segment_index()
-        .expect("Successfully synced above; qed");
+    let target_segment_index = {
+        let last_segment_index = segment_headers_store
+            .max_segment_index()
+            .expect("Successfully synced above; qed");
+
+        if let Some(target_block) = target_block {
+            let mut segment_header = segment_headers_store
+                .get_segment_header(last_segment_index)
+                .ok_or(format!(
+                    "Can't get segment header from the store: {last_segment_index}"
+                ))?;
+
+            if segment_header.last_archived_block().number < target_block {
+                return Err(format!(
+                    "Target block is greater than the last archived block. \
+                    Last segment index = {last_segment_index}, target block = {target_block}, \
+                    last block from the segment = {}
+                    ",
+                    segment_header.last_archived_block().number
+                )
+                .into());
+            }
+
+            let mut current_segment_index = last_segment_index;
+
+            loop {
+                if segment_header.last_archived_block().number == target_block
+                    || current_segment_index <= SegmentIndex::ONE
+                {
+                    break;
+                }
+
+                if segment_header.last_archived_block().number < target_block {
+                    current_segment_index += SegmentIndex::ONE;
+                    break;
+                }
+
+                current_segment_index -= SegmentIndex::ONE;
```

Comment:
Could you maybe use an iterator instead of manually decreasing it in a loop? Something like `(SegmentIndex::ONE..=last_segment_index).rev()` would be much more readable.